scipy/special/_lambertw.h

```diff
@@ -0,0 +1,167 @@
+/* Implementation of the Lambert W function [1]. Based on MPMath
+ *  Implementation [2], and documentation [3].
+ *
+ * Copyright: Yosef Meller, 2009
+ * Authoer email: mellerf@netvision.net.il
+ *
+ * Distributed under the same license as SciPy
+ * Translated into C++ by SciPy developers, 2023.
+ *
+ * References:
+ * [1] On the Lambert W function, Adv. Comp. Math. 5 (1996) 329-359,
+ *     available online: http://www.apmaths.uwo.ca/~djeffrey/Offprints/W-adv-cm.pdf
+ * [2] mpmath source code, Subversion revision 990
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/functions.py?spec=svn994&r=992
+ * [3] mpmath source code, Subversion revision 994
+ *     http://code.google.com/p/mpmath/source/browse/trunk/mpmath/function_docs.py?spec=svn994&r=994
+
+ * TODO: use a series expansion when extremely close to the branch point
+ * at `-1/e` and make sure that the proper branch is chosen there.
+ */
+
+
+#ifndef LAMBERTW_H
+#define LAMBERTW_H
+
+#include <cmath>
+#include <complex>
+#include <numpy/npy_math.h>
+
+#include "_evalpoly.h"
+#include "sf_error.h"
+
+constexpr double EXPN1 = 0.36787944117144232159553;  // exp(-1)
+constexpr double OMEGA = 0.56714329040978387299997;  // W(1, 0)
+
+std::complex<double> lambertw_branchpt(const std::complex<double> z);
+std::complex<double> lambertw_pade0(const std::complex<double> z);
+std::complex<double> lambertw_asy(std::complex<double> z, long k);
+
+
+inline std::complex<double> lambertw(std::complex<double> z, long k, double tol) {
+    int i;
+    double absz;
+    std::complex<double> w;
+    std::complex<double> ew, wew, wewz, wn;
+
+    if (std::isnan(z.real()) || std::isnan(z.imag())) {
+	return z;
+    } else if (z.real() == INFINITY) {
```

Comment:
For constants like `INFINITY`, might be better to use `std::numeric_limits<double>::infinity()`. See [numeric limits](https://en.cppreference.com/w/cpp/types/numeric_limits).